lib/transition/import/cloudfront_access_log_parser.rb

```diff
@@ -0,0 +1,60 @@
+require "ip"
+require "time"
+
+module Transition
+  module Import
+    class CloudfrontAccessLogParser
+      def self.fields
+        @fields || %i[date time x-edge-location sc-bytes c-ip cs-method cs(Host) cs-uri-stem sc-status cs(Referer) cs(User-Agent) cs-uri-query cs(Cookie) x-edge-result-type x-edge-request-id x-host-header cs-protocol cs-bytes time-taken x-forwarded-for ssl-protocol ssl-cipher x-edge-response-result-type cs-protocol-version fle-status fle-encrypted-fields c-port time-to-first-byte x-edge-detailed-result-type sc-content-type sc-content-len sc-range-start sc-range-end]
+      end
+      
+      class Entry < OpenStruct
+        def self.from_string(line)
+          mapping = {}
+          x = nil
+          line.scrub.split(/^(\d{4}-\d{2}-\d{2})\s(\d{2}:\d{2}:\d{2})\s(\w+-?\w+)\s(\d+)\s(\S+)\s(\w+)\s(\S+)\s(\S+)\s(\d+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\w+)\s(\S+)\s(\S+)\s(\w+)\s(\d+)\s(\S+)\s(\S+|-)\s(\S+)\s(\S+)\s(\w+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)\s(\S+)$/).each_with_index do |value, i|
```

Comment:
Hoo boy! I wonder if this might be one of those rare occasions it's worth using the extended flag to allow this beast to be broken up and commented:

```ruby
Regexp.new("
    ^(\d{4}-\d{2}-\d{2})\s # date
    (\d{2}:\d{2}:\d{2})\s # time
    (\w+-?\w+)\s # x-edge-location
    …
  ",
  Regexp::EXTENDED,
)
```

A (more abstract and thus possibly harder to reason about) alternative would be to extract out these regex fragments into the list of fields:

```ruby
CLOUDFRONT_LOG_FIELDS = {
  date: "\d{4}-\d{2}-\d{2}",
  time: "\d{2}:\d{2}:\d{2}",
  :"x-edge-location": "\w+-?\w+",
  …
}

…

def line_regex
  Regexp.new("^" + CLOUDFRONT_LOG_FIELDS.join("\s") + "$")
end
```

One advantage of this approach is that it makes it much less error prone to add or remove a field. If this format is stable, that may not be a concern.

Reply:
Yeah I like these approaches, I will have a play around